feat(linter): Eslint no barrel imports files rule

PrzemekTeclaw opened the pull request:
## Current Behavior
There is no rule checks imports from the same app or lib

## Expected Behavior
"No barrel imports files" rule should disallow imports file from the same app/lib by using alias of current app/lib.
There is configurable option in rule that allows to disallow imports from the index.ts files from the parrent or current root.

PrzemekTeclaw committed:
eslint: no barrel files import rule has been added

PrzemekTeclaw committed:
eslint: unt tests for no barrel files import rule have been added

PrzemekTeclaw committed:
unit tests for no barrel files import rule have been added

PrzemekTeclaw committed:
cleanup(eslint-plugin-nx): no barrel files import rule

export part has been removed from tsConfig interface

PrzemekTeclaw committed:
cleanup(repo): format

code has been formatted

pf1gura commented on packages/workspace/src/schematics/init/init.ts:
```diff
@@ -309,6 +309,12 @@ function updateTsLint() {
         depConstraints: [{ sourceTag: '*', onlyDependOnLibsWithTags: ['*'] }],
       },
     ];
+    tslintJson.rules['no-barrel-files-import'] = [
+      true,
+      {
+        disableImportFromParentAndCurrentRoot: false,
+      },
+    ];
```
Does this rule work with tslint as wel?

pf1gura commented on packages/workspace/src/utils/lint.ts:
```diff
@@ -263,6 +269,12 @@ const globalESLint = JSON.stringify({
+          {
+            disableImportFromParentAndCurrentRoot: false,
+          },
+        ],
```
This part will add this rule for new projects, right?

PrzemekTeclaw replied:
From what I understand - yes.

pf1gura commented on packages/linter/src/migrations/update-10-4-0/update-eslint-configs-to-use-nx-presets.spec.ts:
```diff
@@ -86,6 +86,12 @@ describe('Update ESLint config files to use preset configs which eslint-plugin-n
               ],
             },
           ],
+          '@nrwl/nx/no-barrel-files-import': [
+            'error',
+            {
+              disableImportFromParentAndCurrentRoot: false,
+            },
+          ],
```
Hmm. This is test for migration for future release. Does this mean this migration will enable this rule for existing projects? For existing projects, I think enabling this rule will be a breaking change, so I don't think we should enable it by default.

PrzemekTeclaw committed:
fix(linter): import from folder path in no barrel files import rule

import from folder path has been fixed

PrzemekTeclaw committed:
cleanup(linter): no-barrel-import rule removed

no barrel import rule code has been removed from update-eslint-builder and config specs in migrations

PrzemekTeclaw committed:
cleanup(linter): linter's unit tests files format

linter's unit tests files have been formatted

PrzemekTeclaw committed:
feat(linter): new disallowed import in no barrel files import rule

import from current root has been disallowed in no barrel files import rule
